python/lsst_versions/_versions.py

```diff
@@ -0,0 +1,246 @@
+    for major_release in sorted(releases, reverse=True):
+        if not repo.is_ancestor(commit, releases[major_release]):
+            relevant_release = major_release
+            break
+        if commit == releases[major_release]:
+            # It may be the actual release that should not be a dev
+            # release at all, but for now we do not try to look for
+            # release (non-rc) tags to deal with this properly.
+            relevant_release = major_release - 1
+            break
+
+    if relevant_release == 0:
+        log.warning("Could not find release tag in repo '%s', using 0.", repo_dir)
+
+    # Look through the parents until we find a weekly commit.
+    # The counter can report confusing results if this is being used for
+    # an unmerged development branch.
+    weekly_name, counter = _find_weekly_parent(0, commit, weeklies)
+    if not weekly_name:
+        # No weekly was found. This must be a very early commit.
+        year, week = "0", "0"
+    else:
+        year, week = weekly_name[2:].split(".")
+
+    # Python pre-release versions must can only have a single integer
+    # after the "a".
+    dev_version = f"{relevant_release + 1}.0.0a{int(year):04d}{int(week):02d}{counter:02d}"
+    log.debug(
+        "Using version %s for commit %s derived from weekly %s", dev_version, commit.hexsha, weekly_name
+    )
+
+    return dev_version
+
+
+def _load_toml(content: str) -> Dict[str, Any]:
+    """Parse the TOML contents using deferred import."""
+    import tomli
+
+    return tomli.loads(content)
+
+
+def _write_version(version: str, version_path: str) -> None:
+    """Write the version information to the specified file."""
+    with open(version_path, "w") as fh:
+        print(
+            f"""__all__ = ("__version__",)
+__version__ = "{version}"
+""",
+            file=fh,
+            end="",
+        )
+
+
+def infer_version_for_setuptools(dist: setuptools.Distribution) -> None:
+    """Infer the version and write to the configuration location.
+
+    This function should have been registered as a
+    ``setuptools.finalize_distribution_options`` entry point.
+
+    Parameters
+    ----------
+    dist : `setuptools.Distribution`
+        The setuptools distribution object triggering this code. It will
+        be updated to store the calculated version.
+
+    Notes
+    -----
+    Will look for an entry in the local ``pyproject.toml`` file
+    named ``tool.lsst_versions`` and the key ``write_to`` should
+    be used to specify where the version information should be written.
+
+    Will do nothing if no TOML file can be found.
+    """
+    if not os.path.isfile("pyproject.toml"):
+        return
+
+    with open("pyproject.toml") as fh:
+        parsed = _load_toml(fh.read())
+
+    try:
+        tool = parsed["tool"]["lsst_versions"]
+    except KeyError:
+        # No valid tool entry so nothing to do.
+        return
+
+    write_to = tool.get("write_to")
+    if not write_to:
```

Comment:
Should I issue a warnings.warn message here? It indicates that they did not add a write_to section even though they added a tool section. Seems like that might be something that is a typo.

Reply:
Yup, that sounds kind of iffy.